File: typescript-examples/quick-recipes/api/scrape-without-selectors.ts
import { BrowserContext, Page } from "playwright";
import { extractStructuredData } from "@intuned/browser/ai";
import { z } from "zod";

// Define the schema for a single product
const ProductSchema = z.object({
  name: z.string().describe("Product name"),
  price: z.string().describe("Product price"),
  stock: z.string().describe("Stock status"),
  category: z.string().describe("Product category"),
});

// Define the schema for the list of products
const ProductsSchema = z.object({
  products: z.array(ProductSchema).describe("List of products from the table"),
});

export default async function handler(
  params: any,
  page: Page,
  context: BrowserContext
) {
  await page.goto("https://www.scrapingcourse.com/table-parsing");

  // Extract products using AI - no selectors needed
  const result = await extractStructuredData({
    source: page,
    dataSchema: ProductsSchema,
    prompt: "Extract all products from the table",
  });

  console.log(`Extracted ${result.products.length} products`);
  return result.products;
}

File: python-examples/quick-recipes/api/scrape-without-selectors.py
from playwright.async_api import Page
from pydantic import BaseModel, Field
from intuned_browser.ai import extract_structured_data
from typing import Any, Dict, List


class Product(BaseModel):
    """Schema for a single product."""

    name: str = Field(description="Product name")
    price: str = Field(description="Product price")
    stock: str = Field(description="Stock status")
    category: str = Field(description="Product category")


class Products(BaseModel):
    """Schema for the list of products."""

    products: List[Product] = Field(description="List of products from the table")


async def automation(page: Page, params: Dict[str, Any] | None = None, **_kwargs):
    await page.goto("https://www.scrapingcourse.com/table-parsing")

    # Extract products using AI - no selectors needed
    result = await extract_structured_data(
        source=page,
        data_schema=Products,
        prompt="Extract all products from the table",
    )

    print(f"Extracted {len(result['products'])} products")
    return result["products"]
